fix: Fix batching logic with write records, introduce concurrent requests

nirmeshk opened the pull request:
### Required for all PRs:

- [x] Associated README.md updated.
- [x] Has appropriate unit tests.
- [X] signed CLA

Jagularr commented on plugins/outputs/timestream/timestream.go:
```diff
@@ -269,17 +337,46 @@ func (t *Timestream) Description() string {
 
 func init() {
 	outputs.Add("timestream", func() telegraf.Output {
-		return &Timestream{}
+		t := &Timestream{}
+		return t
```
this change doesn't seem necessary?

nirmeshk replied:
Discarded the change

Jagularr reviewed (commented):
Looks like the test file needs to be formatted

nirmeshk commented:
fixed the gofmt issue with timestream_test.go

ssoroka commented on plugins/outputs/timestream/timestream.go:
```diff
@@ -373,35 +470,36 @@ func (t *Timestream) createTable(tableName *string) error {
 		if len(records) == 0 {
 			continue
 		}
-		id := hashFromMetricTimeNameTagKeys(m)
```
looks like you can also remove the hashFromMetricTimeNameTagKeys function. I'm assuming grouping them by tag and timestamp was not helpful.

nirmeshk replied:
Removed

popey replied:
Sounds like a plan!

nirmeshk replied:
Removed in the latest commit

ssoroka reviewed (changes requested):
change looks thorough, but it's a significant one. Not sure about the parallelism.

nirmeshk commented on plugins/outputs/timestream/timestream.go:
```diff
@@ -275,11 +343,39 @@ func init() {
-		if err := t.writeToTimestream(writeRecordsInput, true); err != nil {
-			return err
-		}
+		go func(inp *timestreamwrite.WriteRecordsInput) {
```
You are right. Currently it is unbounded. 
I can add a semaphore that puts an upper bound on the concurrent go-routines. 
https://github.com/golang/sync/blob/master/semaphore/semaphore.go

We are making this change as we observed that metrics were being dropped due to the requests taking longer serially. After making this change, things improved, and the metric drop stopped. 

Thoughts?

popey replied:
Apologies for the delay in replying to this, we're catching up a little with outstanding PRs. 
Yup, that sounds like a good idea to put an upper bound on the requests, please proceed with that.

nirmeshk replied:
Added the upper bound on the concurrency, and have introduced a parameter for the same. So customers should be able to decide on the concurrency.

nirmeshk commented on plugins/outputs/timestream/timestream.go:
```diff
@@ -275,11 +343,39 @@ func init() {
+
+	// On partial failures, Telegraf will reject the entire batch of metrics and
+	// retry. writeToTimestream will return retryable exceptions only.
+	err, _ := <-errs
```
On line 355, we are only adding to errs channel if err != nil
Reading it n times currently just blocks forever when error does not happen.

I can try removing the != nil check and see if it works un-interupted

popey replied:
Yes, that sounds reasonable, to remove the nil check and always return the result from writeToTimestream.

nirmeshk replied:
I have added a range over channel. It exists on the first encountered error. But since the channel is already closed before that, it should be able to garbage collected?

nirmeshk commented:
Ping on the pull request @Jagularr @ssoroka 
Let me know if the proposal in the comments make sense. I can make the changes and update this

nirmeshk commented:
!signed-cla

hyandell commented:
Nirmesh - CCLA wise, I believe you're pending the Influx folk updating a database on their side.
Influx folk - let me know if you didn't receive the update.

sjwang90 commented:
!signed-cla

powersj commented:
@nirmeshk per #8848 it sounds like this PR improves the situation with the Timestream output. Are you going to be able to resolve some of the outstanding review questions, specifically around limiting the number of threads that get called? Thanks!

nirmeshk commented:
Hi sorry for the delay in addressing this. Picking it up again now. Will raise the request with everything addressed.

nirmeshk commented:
Hi @sjwang90 @powersj @popey , Have addressed all the comments, and have done another round of load testing internally  to make sure there are no concurrency bugs. If you folks can take a look, I would appreciate it.

sysadmin1139 commented:
I was able to use the generated artifacts and verify that our blocking problem has gone away.

sysadmin1139 commented:
I've spent some time today testing the artifacts, and they **clearly** perform better than the existing Telegraf binaries. I haven't thrown full prod loading at it, but it's now surviving our pre-prod throughput tests which version 1.21.1 manifestly doesn't. In a side-by-side test with another metrics platform we are getting highly similar results, so we don't seem to be leaking metrics.

Looking forward to this getting merged.

powersj commented:
> Hi @sjwang90 @powersj @popey , Have addressed all the comments, and have done another round of load testing internally to make sure there are no concurrency bugs. If you folks can take a look, I would appreciate it.

Fix the linter issues and I'm happy to approve.

nirmeshk commented:
@powersj Fixed the lint error. All checks passing now? Have re-based it over the latest changes as well.
